ws/hub.go

```diff
@@ -67,10 +96,73 @@ func (h *Hub) run() {
 	}
 }
 
-// encodeRecord base64 encodes a thread record.
-func encodeRecord(rec thread.Record) []byte {
-	data := rec.RawData()
+// jsonThreadInfo is thread info in JSON.
+type jsonThreadInfo struct {
+	ID        string   `json:"id"`
+	Logs      []string `json:"logs"`
+	FollowKey string   `json:"follow_key"`
+	ReadKey   string   `json:"read_key,omitempty"`
+}
+
+// threadInfoToJSON returns a JSON version of thread info for transport.
+func threadInfoToJSON(info thread.Info) *jsonThreadInfo {
+	var logs []string
+	for _, lg := range info.Logs {
+		logs = append(logs, lg.String())
+	}
+	return &jsonThreadInfo{
+		ID:        info.ID.String(),
+		Logs:      logs,
+		FollowKey: base58.Encode(info.FollowKey.Bytes()),
+		ReadKey:   base58.Encode(info.FollowKey.Bytes()),
+	}
+}
+
+// jsonRecord is a thread record containing link data in JSON.
+type jsonRecord struct {
+	ID       string `json:"id"`
+	LogID    string `json:"log_id,omitempty"`
+	ThreadID string `json:"thread_id"`
+
+	RecordNode string `json:"record_node,omitempty"`
+	EventNode  string `json:"event_node,omitempty"`
+	HeaderNode string `json:"header_node,omitempty"`
+	BodyNode   string `json:"body_node,omitempty"`
+}
+
+// recordToJSON returns a JSON version of a record for transport.
+// Nodes are sent encrypted.
+func recordToJSON(ctx context.Context, dag format.DAGService, rec thread.Record) (*jsonRecord, error) {
+	block, err := rec.GetBlock(ctx, dag)
+	if err != nil {
+		return nil, err
+	}
+	event, err := cbor.EventFromNode(block)
+	if err != nil {
+		return nil, err
+	}
+	header, err := event.GetHeader(ctx, dag, nil)
+	if err != nil {
+		return nil, err
+	}
+	body, err := event.GetBody(ctx, dag, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &jsonRecord{
+		ID:         rec.Cid().String(),
+		RecordNode: encodeNode(rec),
+		EventNode:  encodeNode(block),
+		HeaderNode: encodeNode(header),
+		BodyNode:   encodeNode(body),
+	}, nil
+}
+
+// encodeNode returns a base64-encoded version of n.
+func encodeNode(n format.Node) string {
+	data := n.RawData()
 	encoded := make([]byte, base64.StdEncoding.EncodedLen(len(data)))
 	base64.StdEncoding.Encode(encoded, data)
-	return encoded
+	return string(encoded)
```

Comment:
Nitpick: this is exactly the implementation of [.EncodeToString()](https://golang.org/src/encoding/base64/base64.go?s=4932:4986#L164) :smile:

Reply:
Oh :) changing!